src/core/search/search.cc

```diff
@@ -56,33 +110,60 @@ struct BasicSearch {
     return casted_ptr;
   }
 
-  vector<DocId> Search(monostate, string_view) {
-    return {};
+  // Collect all index results from F(C[i]) and sort by size
+  template <typename C, typename F> vector<IndexResult> GetSubResults(C&& container, F&& f) {
+    vector<IndexResult> sub_results(container.size());
+    for (size_t i = 0; i < container.size(); i++)
+      sub_results[i] = f(container[i]);
+    return sub_results;
   }
 
-  vector<DocId> Search(const AstTermNode& node, string_view active_field) {
-    // Select active indices: search in all text indices if none is selected
-    vector<TextIndex*> selected_indices;
-    if (active_field.empty())
-      selected_indices = indices->GetAllTextIndices();
-    else
-      selected_indices = {GetIndex<TextIndex>(active_field)};
+  using LogicOp = AstLogicalNode::LogicOp;
+
+  // Efficiently unify multiple sub results with specified logical op
+  IndexResult UnifyResults(vector<IndexResult>&& sub_results, LogicOp op) {
+    if (sub_results.empty())
+      return vector<DocId>{};
 
-    // Unify results from all indices
-    vector<DocId> out, tmp;
-    for (auto* index : selected_indices)
-      UnifyResults(index->Matching(node.term), &out, &tmp);
+    auto f = op == AstLogicalNode::AND ? MergeAND : MergeOR;
 
+    // Unifying from smallest to largest is more efficient.
+    // For AND: the result set only shrinks, so start with the smallest.
+    // For OR: unifying smaller sets first reduces the number of element traversals on average.
+    sort(sub_results.begin(), sub_results.end(), IndexResult::BySize);
+
+    IndexResult out{move(sub_results[0])};
+    for (auto& matched : absl::MakeSpan(sub_results).subspan(1))
+      out = f(move(matched), move(out), &tmp_vec);
     return out;
   }
 
-  vector<DocId> Search(const AstRangeNode& node, string_view active_field) {
+  IndexResult Search(monostate, string_view) {
+    return vector<DocId>{};
+  }
+
+  // "term": access field's text index or unify results from all text indices if no field is set
+  IndexResult Search(const AstTermNode& node, string_view active_field) {
+    if (!active_field.empty()) {
+      auto* index = GetIndex<TextIndex>(active_field);
+      return index->Matching(node.term);
```

Comment:
I can foresee Shahar disliking those elegant implicit conversions, right? 🙂

Reply:
I'm ok with that I guess, do you think wrapping them with `IndexResult` (explicit-izing) would make this code harder to read?